zerver/transaction_tests/test_user_groups.py

```diff
@@ -0,0 +1,155 @@
+import threading
+from typing import TYPE_CHECKING, List, Optional
+
+import orjson
+from django.db import connections, transaction
+from django.utils.timezone import now as timezon_now
+
+from zerver.actions.user_groups import add_subgroups_to_user_group, check_add_user_group
+from zerver.lib.test_classes import ZulipTransactionTestCase
+from zerver.models import Realm, UserGroup, get_realm
+from zerver.views.development import user_groups as user_group_view
+
+if TYPE_CHECKING:
+    from django.test.client import _MonkeyPatchedWSGIResponse as TestHttpResponse
+
+
+class UserGroupRaceConditionTestCase(ZulipTransactionTestCase):
+    created_user_groups: List[UserGroup] = []
+
+    def tearDown(self) -> None:
+        # Clean up the user groups created to minimize leakage
+        with transaction.atomic():
+            for group in self.created_user_groups:
+                group.delete()
+            transaction.on_commit(lambda: self.created_user_groups.clear())
+
+        super().tearDown()
+
+    def create_user_group_chain(self, realm: Realm) -> List[UserGroup]:
+        """Build a user groups forming a chain through group-group memberships
+        returning a list where each group is the supergroup of its subsequent group.
+        """
+        groups = [
+            check_add_user_group(realm, f"chain #{timezon_now()}", [], acting_user=None)
```

Comment:
`timezon_now` has a typo. Also, generally we don't want to use time since it might fail to tick between elements of the loop leading to flakes; unless you're using the time machine module. Can we just have it be the next integer instead? Or maybe a nicer API would just have the caller pass a list of group names or something, and generate a chain using those groups; not sure.